### DISCUSSION TITLE
How to efficiently get IO pins positions and reuse them

### DISCUSSION THREAD

**Discussion by lustefan97** [The-OpenROAD-Project/OpenROAD-flow-scripts#1235] [Q&A] [answered]
I am currently working on my master thesis where the subject is to enable 3D system implementation using OpenROAD, to achieve 3D stacked designs using face-to-face IO bonding.

I am trying to implement a flow (using the Ibex design as a base), and I already got the bottom die working (detailled routing did not finished, but I think that this is due to the use of the skywater PDK, which is a very old node). I want now to go further and work on the implementation of the top die and to do so I first need to get the IO pins positions of the bottom one. 

This is were I am kind of stuck and would like your insight. 
My idea is simply to write a .DEF file for the bottom die and read it to store the name and positions of its IO pins into a list, this was all fine and dandy up until I tried to use the "place_pin" command to manually place the pins knowing their name and positions.

Here is my issue : I read and stored the IO's names and positions using a TCL script, I divided the coordinates by 1000 to convert to micron, but when I tried to use the function I got the following error : 

" TypeError in method 'microns_to_dbu', argument 1 of type 'double' "

So my questions now are : 
1) What does this error mean ?
2) How can I solve this ? 
3) Would there be a more efficient/more reliable way to get the pin positions (and eventually modify them to mirror their positions with regard to the x axis, as right now I am just substracting each x position to the die right side position to do so) ?
4) (Since I am here) Is there a possibility to display 2 dies on the gui simultaneously ? Whether it would be on top of one another, or beside one another.

I admit I converted the coordinates to double before dividing them, but I didn't thought that it would be a problem... Let me know if you need more details

**Comment by vijayank88** [comment]
@lustefan97 
Did you refer this [document](https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/ppl)?

**Reply by lustefan97** [reply]
Yes, and I've successfully placed I/O pins on the top layer of the (bottom) die. The thing now is that I need to use their coordinates (that was decided by the tool) for the die that will be stacked on top, and it needs to be automatized since I cannot enter manually the position of each pin (there are 264 pins in the ibex design, but I need to go as high as 10 000 if not 100 000 pins). 

So as of right now, I have gathered the (already placed) pins name and coordinates in a list using a tcl script but if I try to use the place_pin command by passing each x and y coordinates (in micron) as a variable (for example : $xPos = 428.6 and $yPos = 428.6) then I encounter the error that I mentionned but if I enter those number manually, no issue... 

From what I understood TCL is not typed, so I don't get what the error is about (maybe that I want to pass a double where it needs a string, but again I haven't found anything about this error in the documentation or previous threads...)

Also I feel that there may be a more efficient way to get the pins coordinates than creating an extraneous file and reading it afterward, but again I didn't read anything about something in that vein in the documentation. This is why I'm also asking for some insight, to maybe hit two birds in one stone by bypassing the issue and creating something more stable

**Reply by vijayank88** [reply]
Please write into `.def` format and check co-ordinates. This will give exact location for each pins.
@maliberty Is there other way to dump the locations of pins?

**Reply by lustefan97** [reply]
I know, that is exactly what I did and I've got them in a list with the name of the associated pin. 

My issue is with the "reuse" part, I cannot use the coordinates without encountering the TypeError message and I don't want to enter each coordinates manually. 
I wrote a script reading the  pins names and coordinates and putting them in a list to reuse in another design, but when I want to use the place_pin command with those stored coordinates I have the mentionned error on the "micron_to_db" method, and I don't get why or what it means since I have read nothing about it in the documentation.

**Reply by vijayank88** [reply]
Are you using the latest commit?
Is that possible to provide reproducible test case?

**Reply by vijayank88** [reply]
@lustefan97 
`ord::dbu_to_microns <value>` to convert dbu to microns and use that value to place the pins.

**Reply by lustefan97** [reply]
I think I tried that method before in my tcl script, but I may have used it wrongly as it caused the application to crash...

I am unable to try it again right now, but I will as soon as I am able to and I'll let you know ! Thank you for your help !

**Comment by maliberty** [comment]
Can you show exactly the TCL you are executing that causes the type error?

**Reply by lustefan97** [reply]
Yes sure, here is the part of the script were I am extracting the pin positions :

<img width="471" alt="pinExtraction1" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/ec594af5-5093-4329-8dc1-810b558df144">

Then here is the command that I want to use, and what I do (it really is the place_pin that gives "TypeError in method 'microns_to_dbu', argument 1 of type 'double' ") : 

<img width="502" alt="pinExtraction2" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/a9e47c2f-859e-4061-bd12-36127b108519">

And this is what I wanted to write at first : 


<img width="689" alt="pinExtraction3" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/b11db054-d1b2-4011-b93f-9d875b433e3b">

Also I tried to use "ord::dbu_to_microns" as recommended previously, and as used in the save_image.tcl script (same syntax), in the gui TCL command box, but it unfortunately make the gui crash... Just for full disclosure, I base myself off of the flow.tcl script found in OpenROAD-flow-scripts/tools/OpenROAD/test and test my scripts there, if it is a problem or not recommended, please let me know !

**Reply by maliberty** [reply]
What are the actual values of xPos and yPos when it fails?

**Reply by lustefan97** [reply]
With the values I get with the 2nd screenshot (IO Placement), I have $xPos = 241.2 and $yPos = 496.2

<img width="98" alt="pins1" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/d1d1722d-cf56-4b57-87b4-39bd9f3efa1c">

But I also tried with integers like with $xPos =  241 and $yPos = 496 by changing %.1f to %d in "format", and I have the same error "microns_to_dbu" that is raised

<img width="83" alt="pins2" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/c4ce5c78-9cfc-402d-9aac-194c4acb39da">

**Comment by rovinski** [comment]
@lustefan97 there is no reason to be parsing the DEF to get the pin locations. You want to be using the DB interface. You can get a list of the BTerm objects with

```
set bterms [[[[ord::get_db] getChip] getBlock] getBTerms]
```

From there you can extract whatever properties you need. For example if you want the bounding box, you could use a script like this
```
foreach bterm $bterms {
  set bbox [$bterm getBBox]
  puts "Name: [$bterm getName]. BBox: [$bbox ll] [$bbox ur]"
}
```

The BBox points are reported in DBU so you can use `ord::dbu_to_microns` to convert to micron if that's what you need.

You should be able to take all this information and create a script which you can then use with  `place_pin` to place the pins in the exact locations you need.

**Reply by lustefan97** [reply]
I tried to use `ord::dbu_to_microns` to convert the coordinates in micron, but the gui crashed due to a segmentation fault.

Here is the simple test I tried, I also tried with the real pin positions with the same result (any other `ord::` method call cause this too) : 

<img width="492" alt="dbu_to_micron test" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/0937ddf4-0609-4c81-b325-7277db713430">

Here is the report I got : 

<img width="836" alt="dbu_to_micron crash message" src="https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/61243950/92906ca9-a6aa-4e72-815a-3e9f84e36868">

I use Ubuntu 22.04 on a Windows Subsystem for Linux, I had to re-install OpenROAD fairly recently (maybe a month a half ago) is it an issue that was resolved recently or could it be due to the use of WSL ?

**Reply by maliberty** [reply]
If you haven't loaded a LEF then the conversion is meaningless.  We should not crash though and can give an error instead.

**Reply by maliberty** [reply]
The crash is addressed in https://github.com/The-OpenROAD-Project/OpenROAD/pull/3652

**Reply by lustefan97** [reply]
@maliberty @rovinski 

Thank you so much for your help, I'll definitely use those methods from now on ! 
It answers a part of my question, but I still have a TypeError due to `ord::microns_to_dbu` that comes on when I want to use place_pin and even now with use the DB interface and `ord::dbu_to_microns` (and by curiosity I checked what happens when I don't convert to microns, and the result is the same). 

I don't know exactly what this error means, and how I can solve it since from my understanding TCL is not a typed language and basically everything can be considered as a string (and if true, to me, that excludes a theory I had that place_pin expects a string instead of the double that is passed to it since the number is stored as a double in memory). 

I have here a small script that I used to quickly test placing pins, get their positions and use place_pin to place them back a the exact same place (the expected behaviour being that nothing changes, or I get a warning/error saying that something has already been placed there). I use it in the TCL commands box of the gui while in _OpenROAD-flow-scripts/tools/OpenROAD/test_, if you want to try for yourself. The exact error I get is   **[ERROR GUI-0070] TypeError in method 'microns_to_dbu', argument 1 of type 'double'** 


``` 
source "helpers.tcl"
source "flow_helpers.tcl"

source "sky130hd/sky130hd.vars"
set place_density 0.7   

set design "ibex"
set top_module "ibex_core"
set synth_verilog "ibex_sky130hd.v"
set sdc_file "ibex_sky130hd.sdc"
set core_utilization 45

set slew_margin 30
set cap_margin 25

read_libraries
read_verilog $synth_verilog
link_design $top_module
read_sdc $sdc_file

# Step 1 : Initialize the floorplan of the die
#-----------------------------------------------
set aspect_ratio 1.0
set core_margin 1.0

initialize_floorplan -utilization $core_utilization \
                     -aspect_ratio $aspect_ratio    \
                     -core_space $core_margin       \
                     -site $site                                 

# Source the tracks (for : ) definition  
source $tracks_file

# Remove the buffers introduced by the synthesis tool (during synthesis).
remove_buffers   


define_pin_shape_pattern -layer met5 -x_step 3.4 -y_step 3.4 -region {10 10 525 525} -size {1.6 1.6}
set_io_pin_constraint -pin_names * -region "up:*"
place_pins -random -hor_layers $io_placer_hor_layer -ver_layers $io_placer_ver_layer

eval tapcell $tapcell_args

source $pdn_cfg
pdngen

global_placement -skip_io -density $place_density -pad_left $global_place_pad \
                                                  -pad_right $global_place_pad

define_pin_shape_pattern -layer met5 -x_step 3.4 -y_step 3.4 -region {10 10 525 525} -size {1.6 1.6}
place_pins -hor_layer $io_placer_hor_layer -ver_layer $io_placer_ver_layer


#------------------------------------------------------------------------------
# This is where I am trying to get the pins coordinates and use place_pin 
#------------------------------------------------------------------------------

# Get the pins locations info + set half pin width value to add to the low left coordinates

set bterms [[[[ord::get_db] getChip] getBlock] getBTerms]
set hpinw 0.8

foreach bterm $bterms {
  set bbox [$bterm getBBox]
  puts "Name: [$bterm getName]. BBox: [$bbox ll] [$bbox ur]"
  puts "Name of the pin : [$bterm getName]"
  #set one [lindex [$bbox ll] 0]
  set one [ord::dbu_to_microns [lindex [$bbox ll] 0]]  
  set one [expr {$one + $hpinw}]
  puts "X position : $one"
  #set two [lindex [$bbox ll] 1]
  set two [ord::dbu_to_microns [lindex [$bbox ll] 1]]
  set two [expr {$two + $hpinw}]
  puts "Y position : $two"
  place_pin -pin_name [$bterm getName] -layer met5 -location {$one $two} -pin_size {1.6 1.6}
}

# This last line with place_pin is what is giving "TypeError" 
```

**Reply by rovinski** [reply, marked as answer]
@lustefan97 You're getting caught by Tcl syntax. Braces (`{}`) create literal expressions and don't allow dereferencing variables with `$`.
When you use `-location {$one $two}` it resolves to a list of the two strings `$one $two` which are indeed not convertible to doubles.

Change the line to this and it should work:
```
place_pin -pin_name [$bterm getName] -layer met5 -location [list $one $two] -pin_size {1.6 1.6}
```

**Reply by rovinski** [reply]
Another thing is that you can look up methods of objects between [`db.h`](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/include/odb/db.h) and [`geom.h`](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/include/odb/db.h). There are better accessor methods for `Rect` objects.

`[lindex [$bbox ll] 0]]` is equivalent to `[$bbox xMin]`
Instead of getting the lower left plus some offset you could just use `$bbox xCenter` and `$bbox yCenter`
etc.

You can also mistype a method for an object in the shell to print out a list of methods. It's sometimes faster.

```
openroad> $bbox asdf
Invalid method. Must be one of: configure cget -acquire -disown -delete = init reset moveTo moveDelta mergeInit isInverted minDXDY maxDXDY getDir set_xlo set_xhi set_ylo set_yhi xMin yMin xMax yMax dx dy xCenter yCenter getPoints ll ul ur lr intersects overlaps contains inside closestPtInside merge bloat intersection intersect area margin printf print
```

**Reply by lustefan97** [reply]
I see, I tried a similar line but always with braces and switching to brackets solved the problem, some TCL trickiness I guess. I'll keep that in mind from now on, thanks for your help !
